File: flutter/contracts/lib/src/module_event.dart
/// Base class for all cross-module events.
///
/// All events must extend [ModuleEvent] for type-safe subscription.
/// Mirrors Android sealed class [ModuleEvent].
///
/// Usage:
/// ```dart
/// context.eventBus.on<OrderCreatedEvent>((e) => refresh());
/// context.eventBus.publish(OrderCreatedEvent(orderId: '123', moduleId: 'orders'));
/// ```
abstract class ModuleEvent {
  final DateTime timestamp;
  ModuleEvent() : timestamp = DateTime.now();
}

// ── Auth events ─────────────────────────────────────────────────────────────

class UserAuthenticatedEvent extends ModuleEvent {
  final String userId;
  final String role; // Role enum value name
  UserAuthenticatedEvent({required this.userId, required this.role});
}

class UserLoggedOutEvent extends ModuleEvent {}

// ── Module lifecycle events ──────────────────────────────────────────────────

class ModuleInitializedEvent extends ModuleEvent {
  final String moduleId;
  ModuleInitializedEvent({required this.moduleId});
}

// ── Domain events ────────────────────────────────────────────────────────────

class OrderCreatedEvent extends ModuleEvent {
  final String orderId;
  final String moduleId;
  OrderCreatedEvent({required this.orderId, required this.moduleId});
}

// ── Tenant events ────────────────────────────────────────────────────────────

class TenantSwitchedEvent extends ModuleEvent {
  final String oldTenantId;
  final String newTenantId;
  TenantSwitchedEvent({required this.oldTenantId, required this.newTenantId});
}

class TenantConfigUpdatedEvent extends ModuleEvent {
  final String tenantId;
  TenantConfigUpdatedEvent({required this.tenantId});
}

// ── Feature flag events ──────────────────────────────────────────────────────

class FeatureFlagChangedEvent extends ModuleEvent {
  final String key;
  final bool enabled;
  FeatureFlagChangedEvent({required this.key, required this.enabled});
}

// ── Navigation events ────────────────────────────────────────────────────────

class NavigationRequestedEvent extends ModuleEvent {
  final String route;
  final Map<String, String> args;
  NavigationRequestedEvent({required this.route, this.args = const {}});
}

File: ios/Packages/Contracts/Sources/Contracts/ModuleEvent.swift
import Foundation

/// Base class for all cross-module events.
///
/// All events must inherit from `ModuleEvent` for type-safe subscription.
/// Mirrors Android sealed class `ModuleEvent`.
///
/// Usage:
/// ```swift
/// eventBus.on(OrderCreatedEvent.self)
///     .sink { event in refresh() }
///     .store(in: &cancellables)
///
/// eventBus.publish(OrderCreatedEvent(orderId: "123", moduleId: "orders"))
/// ```
open class ModuleEvent {
    public let timestamp: Date
    public init() { self.timestamp = Date() }
}

// ── Auth events ──────────────────────────────────────────────────────────────

public final class UserAuthenticatedEvent: ModuleEvent {
    public let userId: String
    public let role: Role
    public init(userId: String, role: Role) {
        self.userId = userId
        self.role = role
        super.init()
    }
}

public final class UserLoggedOutEvent: ModuleEvent {
    public override init() { super.init() }
}

// ── Module lifecycle events ──────────────────────────────────────────────────

public final class ModuleInitializedEvent: ModuleEvent {
    public let moduleId: String
    public init(moduleId: String) {
        self.moduleId = moduleId
        super.init()
    }
}

// ── Domain events ────────────────────────────────────────────────────────────

public final class OrderCreatedEvent: ModuleEvent {
    public let orderId: String
    public let moduleId: String
    public init(orderId: String, moduleId: String) {
        self.orderId = orderId
        self.moduleId = moduleId
        super.init()
    }
}

// ── Tenant events ────────────────────────────────────────────────────────────

public final class TenantSwitchedEvent: ModuleEvent {
    public let oldTenantId: String
    public let newTenantId: String
    public init(oldTenantId: String, newTenantId: String) {
        self.oldTenantId = oldTenantId
        self.newTenantId = newTenantId
        super.init()
    }
}

public final class TenantConfigUpdatedEvent: ModuleEvent {
    public let tenantId: String
    public init(tenantId: String) {
        self.tenantId = tenantId
        super.init()
    }
}

// ── Feature flag events ──────────────────────────────────────────────────────

public final class FeatureFlagChangedEvent: ModuleEvent {
    public let key: String
    public let enabled: Bool
    public init(key: String, enabled: Bool) {
        self.key = key
        self.enabled = enabled
        super.init()
    }
}

// ── Navigation events ────────────────────────────────────────────────────────

public final class NavigationRequestedEvent: ModuleEvent {
    public let route: String
    public let args: [String: String]
    public init(route: String, args: [String: String] = [:]) {
        self.route = route
        self.args = args
        super.init()
    }
}
